File: web/src/services/alertEvent.ts
import { request } from './api';

// Alert Event types
export interface AlertEvent {
  id: string;
  rule_id: string;
  node_id: string;
  metric_value: number;
  status: string; // firing, resolved, acknowledged
  fired_at: string;
  resolved_at?: string;
  acknowledged_by?: string;
  acknowledged_at?: string;
  notify_sent: boolean;
  notify_sent_at?: string;

  // Joined fields
  rule_name?: string;
  node_name?: string;
  metric_type?: string;
  operator?: string;
  threshold?: number;
}

export interface AlertEventListParams {
  page?: number;
  page_size?: number;
  node_id?: string;
  rule_id?: string;
  status?: string;
  from?: string;
  to?: string;
  [key: string]: string | number | undefined;
}

export interface AcknowledgeAlertEventRequest {
  acknowledged_by: string;
}

function buildQuery(params: Record<string, string | number | undefined>): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== '') sp.set(k, String(v));
  }
  return sp.toString() ? `?${sp}` : '';
}

export const alertEventApi = {
  list: (params?: AlertEventListParams) => {
    const query = buildQuery(params || {});
    return request<{ list: AlertEvent[]; total: number; page: number; page_size: number }>(`/alert-events${query}`);
  },
  acknowledge: (id: string, acknowledgedBy: string) =>
    request<void>(`/alert-events/${id}/acknowledge`, {
      method: 'POST',
      body: JSON.stringify({ acknowledged_by: acknowledgedBy }),
    }),
};

File: app/internal/dto/alert_event.go
package dto

// AlertEventListRequest 告警事件列表查询参数
type AlertEventListRequest struct {
	PageRequest
	NodeID   string `form:"node_id"`
	RuleID   string `form:"rule_id"`
	Status   string `form:"status" binding:"omitempty,oneof=firing resolved acknowledged"`
	From     string `form:"from"` // RFC3339
	To       string `form:"to"`   // RFC3339
}

// AcknowledgeAlertEventRequest 确认告警事件请求
type AcknowledgeAlertEventRequest struct {
	AcknowledgedBy string `json:"acknowledged_by" binding:"required,min=1,max=255"`
}

// AlertEventResponse 告警事件响应
type AlertEventResponse struct {
	ID              string   `json:"id"`
	RuleID          string   `json:"rule_id"`
	NodeID          string   `json:"node_id"`
	MetricValue     float64  `json:"metric_value"`
	Status          string   `json:"status"`
	FiredAt         string   `json:"fired_at"`
	ResolvedAt      *string  `json:"resolved_at,omitempty"`
	AcknowledgedBy  *string  `json:"acknowledged_by,omitempty"`
	AcknowledgedAt  *string  `json:"acknowledged_at,omitempty"`
	NotifySent      bool     `json:"notify_sent"`
	NotifySentAt    *string  `json:"notify_sent_at,omitempty"`

	// Joined fields
	RuleName    string `json:"rule_name,omitempty"`
	NodeName    string `json:"node_name,omitempty"`
	MetricType  string `json:"metric_type,omitempty"`
	Operator    string `json:"operator,omitempty"`
	Threshold   float64 `json:"threshold,omitempty"`
}
